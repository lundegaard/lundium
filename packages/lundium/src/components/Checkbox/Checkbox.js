import React from 'react';
import PropTypes from 'prop-types';

import CheckboxControl from '../CheckboxControl';
import FormField from '../FormField';

const Checkbox = props => <FormField control={CheckboxControl} kind="checkbox" {...props} />;

Checkbox.propTypes = {
	/** Additional content to render after the control. */
	afterControl: PropTypes.element,
	/** Additional content to render before the control. */
	beforeControl: PropTypes.element,
	/** Field's control component */
	control: PropTypes.elementType.isRequired,
	/** If set, control is disabled */
	disabled: PropTypes.bool,
	/** Node shown in error state. */
	error: PropTypes.node,
	formGroupClassName: PropTypes.string,
	/** @ignore */
	generatedID: PropTypes.string,
	/** Additional props to pass to the FormGroup */
	groupProps: PropTypes.object,
	/** If `true`, input label has floating animation. */
	hasFloatingLabel: PropTypes.bool,
	/** Used to pair the control with a label. If undefined, we use a generated one. */
	id: PropTypes.any,
	/** If `true`, label is absolutely positioned on the right side of the form group. */
	inline: PropTypes.bool,
	/** If `true`, label is hidden. */
	isLabelHidden: PropTypes.bool,
	/** Used to pass a context className to FormGroup (to properly style FormControl and Label) */
	kind: PropTypes.string,
	/** Label value */
	label: PropTypes.node,
	/** Additional props to pass to the Label */
	labelProps: PropTypes.object,
	/** Legend inside text input. */
	legendText: PropTypes.string,
	/** Name of the field */
	name: PropTypes.string.isRequired,
	/** If set, control is readOnly */
	readOnly: PropTypes.bool,
	/** If set, control is required */
	required: PropTypes.bool,
	/** Assign a suffix component (e.g. Icon) to Input */
	suffix: PropTypes.node,
	/** Node shown in warning state. */
	warning: PropTypes.node,
};

export default Checkbox;
