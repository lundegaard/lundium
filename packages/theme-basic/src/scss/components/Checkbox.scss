@import '../core';

/*
 -  Checkbox
 -  (<input type="checkbox" class="form-control" />)
 */


.form-group--checkbox {
	position: relative;
	display: block;
	min-height: 20px;
	padding-block-start: 0;
	padding-inline-start: ( $custom-control-gutter + $custom-control-indicator-size );

	.control-label {
		position: relative;
		margin-block-end: 0;
		vertical-align: middle;
		padding-inline-start: 0;
		font-size: $control-label-font-size;

		&::before {
			content: '';
			position: absolute;
			inset-block-start: -( $font-size-base * $line-height-base - $custom-control-indicator-size ) / 2;
			inset-inline-start: -( $custom-control-gutter + $custom-control-indicator-size );
			width: $custom-control-indicator-size;
			height: $custom-control-indicator-size;
			display: block;
			pointer-events: none;
			background-color: $custom-control-indicator-bg;
			border: $custom-control-indicator-border-color solid $custom-control-indicator-border-width;
			box-shadow: $custom-control-indicator-box-shadow;
			border-radius: $custom-checkbox-indicator-border-radius;
			transition: all $transition-speed $transition-easing;
		}

		&::after {
			content: '';
			position: absolute;
			inset-block-start: -( $font-size-base * $line-height-base - $custom-control-indicator-size ) / 2;
			inset-inline-start: -( $custom-control-gutter + $custom-control-indicator-size );
			display: block;
			width: $custom-control-indicator-size;
			height: $custom-control-indicator-size;
			background: no-repeat 50% / 60% 60%;
		}

		&.sr-only {
			width: auto;
			height: auto;
			overflow: visible;
			text-indent: -9999px;
		}
	}

	.form-control {
		position: absolute;
		height: auto;
		width: 1px;
		z-index: -1;
		opacity: 0;

		&:checked ~ .control-label::before {
			color: $custom-control-indicator-checked-color;
			border-color: $custom-control-indicator-checked-border-color;
			background-color: $input-focused-bg;
			box-shadow: $custom-control-indicator-checked-box-shadow;
		}

		&:checked ~ .control-label::after {
			background-image: $custom-checkbox-indicator-icon-checked;
		}

		/* &:indeterminate ~ .control-label {
			&::before {
				border-color: $custom-checkbox-indicator-indeterminate-border-color;
				@include gradient-bg($custom-checkbox-indicator-indeterminate-bg);
				box-shadow: $custom-checkbox-indicator-indeterminate-box-shadow;
			}

			&::after {
				background-image: $custom-checkbox-indicator-icon-indeterminate;
			}
		} */

		&:focus ~ .control-label::before {
			box-shadow: $custom-control-indicator-focus-box-shadow;
		}

		&:focus:not(:checked) ~ .control-label::before {
			border-color: $custom-control-indicator-focus-border-color;
		}

		&:not(:disabled):active ~ .control-label::before {
			color: $custom-control-indicator-active-color;
			background-color: $custom-control-indicator-active-bg;
			border-color: $custom-control-indicator-active-border-color;
			box-shadow: $custom-control-indicator-active-box-shadow;
		}

		&:not(:disabled):not(:read-only) ~ .control-label {
			cursor: pointer;

			&:hover {
				&::before {
					border-color: $input-hover-border-color;
				}
			}
		}

		&:disabled {
			opacity: 0;

			&:checked ~ .control-label::before {
				background-color: $custom-control-indicator-checked-disabled-bg;
			}

			/* &:indeterminate ~ .custom-control-label::before {
				background-color: $custom-control-indicator-checked-disabled-bg;
			} */

			~ .control-label {
				color: $control-label-disabled-color;

				&::before {
					background-color: $custom-control-indicator-disabled-bg;
				}
			}
		}

		&.is-invalid ~ .control-label::before {
			border-color: $danger;
		}
	}
}
