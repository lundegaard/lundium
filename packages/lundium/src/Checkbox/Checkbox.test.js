import React from 'react';
import { shallow } from 'enzyme';

import Checkbox from './Checkbox';

describe('<Checkbox />', () => {
	const requiredProps = {};

	it('matches snapshot when shallowly rendered', () => {
		const wrapper = shallow(<Checkbox {...requiredProps} />);
		expect(wrapper).toMatchSnapshot();
	});
});
