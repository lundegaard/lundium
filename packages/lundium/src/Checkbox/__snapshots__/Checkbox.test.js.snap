// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> matches snapshot when shallowly rendered 1`] = `
<forwardRef(FormField)
  control={[Function]}
  kind="checkbox"
/>
`;
